explorer/interpreter/type_checker.cpp

```diff
@@ -4462,6 +4464,93 @@ auto TypeChecker::TypeCheckCallableDeclaration(Nonnull<CallableDeclaration*> f,
   return Success();
 }
 
+auto TypeChecker::CheckPatternEquality(Nonnull<const Pattern*> o_patt,
+                                       Nonnull<const Pattern*> e_patt)
+    -> ErrorOr<Success> {
+  if (o_patt->kind() != e_patt->kind()) {
+    return ProgramError(o_patt->source_loc())
+           << __FUNCTION__ << ":" << __LINE__ << " failed. kind mismatch";
+  }
+  switch (e_patt->kind()) {
+    case PatternKind::BindingPattern: {
+      const auto* o = cast<BindingPattern>(o_patt);
+      const auto* e = cast<BindingPattern>(e_patt);
+
+      CARBON_RETURN_IF_ERROR(CheckPatternEquality(&o->type(), &e->type()));
+
+      break;
+    }
+    case PatternKind::TuplePattern: {
+      const auto* o = cast<TuplePattern>(o_patt);
+      const auto* e = cast<TuplePattern>(e_patt);
+
+      if (o->fields().size() != e->fields().size()) {
+        return ProgramError(o->source_loc())
+               << __FUNCTION__ << ":" << __LINE__ << " failed. size mismatch";
+      }
+
+      for (auto pair : llvm::zip(o->fields(), e->fields())) {
+        CARBON_RETURN_IF_ERROR(
+            CheckPatternEquality(std::get<0>(pair), std::get<1>(pair)));
+      }
+
+      break;
+    }
+    case PatternKind::ExpressionPattern: {
+      const auto* o = cast<ExpressionPattern>(o_patt);
+      const auto* e = cast<ExpressionPattern>(e_patt);
+
+      CARBON_RETURN_IF_ERROR(
+          CheckExpressionEquality(&o->expression(), &e->expression()));
+      break;
+    }
+    default:
+      if (trace_stream_) {
+        **trace_stream_ << "cradtke: unexpected pattern kind - "
+                        << PatternKindName(e_patt->kind()) << "\n";
+      }
+      //      CARBON_CHECK(0 && "unreachable");
+      break;
+  }
+  return Success();
+}
+
+auto TypeChecker::CheckExpressionEquality(Nonnull<const Expression*> o_exp,
+                                          Nonnull<const Expression*> e_exp)
+    -> ErrorOr<Success> {
+  if (o_exp->kind() != e_exp->kind()) {
+    return ProgramError(o_exp->source_loc())
+           << __FUNCTION__ << ":" << __LINE__ << " failed. kind mismatch";
+  }
```

Comment:
I think this is too strong as it should be legal (I think?) for `o_exp->kind()` to be `IntLiteral` and `e_exp->kind()` to be `IntTypeLiteral`.